### PR TITLE
feat: support for multiple external network

### PR DESCRIPTION
**Pull request by shane965**
- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

### What type of this PR
Examples of user facing changes:
- Features

### Which issue(s) this PR fixes:
Fixes #(issue-number)

### <samp>🤖 Generated by Copilot at 0b62674</samp>

This pull request adds support for configurable external networks for NAT gateways and EIPs in the kube-ovn project. It updates the CRD definitions, the Go structs, and the controller logic to handle the new fields `externalSubnets` and `externalSubnet` for `VpcNatGateway` and `IptablesEIP` respectively. It also removes the dependency on the default external network name and improves the subnet status and EIP allocation calculations.

### <samp>🤖 Generated by Copilot at 0b62674</samp>

> _No more default external network_
> _We choose our own destiny_
> _`externalSubnets` and `externalSubnet`_
> _We unleash the power of NAT and EIP_

### <samp>🤖 Generated by Copilot at 0b62674</samp>

*  Add `externalSubnets` field to `VpcNatSpec` and `externalSubnet` field to `IptablesEipSpec` in CRDs and types ([link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-d671263a0aa28dac78faa045e43d18cde05bf81846c6d1bf1db009cddda883f2R179-R182), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-d671263a0aa28dac78faa045e43d18cde05bf81846c6d1bf1db009cddda883f2R565-R566), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aL486-R492), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-69f75d53361877cc02ab6f413f306486d29287c6400b92b71c4c74f018f3234aL507-R513), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-7b60edb5fc7c18c120f5a0077606604cfb6be9bb55517095bb9380877ef716c5R179-R182), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-7b60edb5fc7c18c120f5a0077606604cfb6be9bb55517095bb9380877ef716c5R565-R566))
* Update `DeepCopyInto` method for `VpcNatSpec` to copy `externalSubnets` field ([link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-06f8e5bc84406eb8a89c72c345fc4ff5f144e480103ad55d681f7567f0f85515R2389-R2393))
* Remove `VpcExternalNet` constant from `const.go` and add it to `net.go` ([link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-23fe1930ce750ec27bfa79a8fb8d076245c64df2b62ea6d6367cdd59260b164cL167), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-a1ffb52e88506b4737c752182fb157473a2575e8e7bb356ed709d2316141a3cdR29-R30))
* Add utility functions `GetExternalNetwork` and `GetNatGwExternalNetwork` to `net.go` to get external network name from user input or default value ([link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-a1ffb52e88506b4737c752182fb157473a2575e8e7bb356ed709d2316141a3cdR512-R527))
* Remove `checkVpcExternalNet` method from `vpc_nat_gateway.go` as it is no longer needed ([link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4L83-L89), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4L766-L777))
* Update `vpc_nat_gateway.go` to use `GetNatGwExternalNetwork` function and `externalSubnets` field to get external network name, gateway, CIDR, and pod annotation for NAT gateway ([link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4L536-R534), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-41b4c692987143b2796e2664174ba6858e7c51329d823c612556ef6cd32185c4L670-R667))
* Remove `MACVLAN_NAD_PROVIDER` constant from `vpc_nat_gw_eip.go` as it is no longer needed ([link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L24-L28))
* Update `vpc_nat_gw_eip.go` to use `GetExternalNetwork` function and `externalSubnet` field to get external network name, EIP CIDR, gateway, static address, and label for EIP ([link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L54-R50), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L64-R63), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L226-R229), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L234-R244), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L258-R259), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L337-R342), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L408-R410), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L414-R416), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L600-R602), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L610-R613), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L617-R627), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L652-R657), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L670-R674), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L679-R682), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L739-R748), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92R979), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L989-R993), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-c2fca6de8b38cf14047d762d1029e969197a5fd0d5172062dce753f1e9613c92L995-R999))
* Update `subnet.go` to check if subnet is OVN subnet and list EIPs by label selector based on external network name ([link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1872-R1872), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-3742843fd11c637b26eae183fb8a0c4eeff96bbfb18a8db1e0ded0c33733f205L1943-R1943))
* Update `init.go` to get static address for EIP from IPAM based on `externalSubnet` field ([link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-92aa62435bc2b1b2be34f478f650597068bcf35b8d5975d119f91cf78ac54988L437-R438))
* Update `pod_iptables_eip.go` to get EIP object from lister and pass `externalNetwork` parameter to `createOrUpdateCrdEip` method ([link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-2d73329e26ec408ad10253e5352451fb5d5db2bac81ebb1ead0e546e8df7e061L255-R255), [link](https://github.com/kubeovn/kube-ovn/pull/2725/files?diff=unified&w=0#diff-2d73329e26ec408ad10253e5352451fb5d5db2bac81ebb1ead0e546e8df7e061L267-R268))

### CODE REVIEW

**Comment by shane965** [comment]
in order to keep compatibility with the original code, we must to use the same function to get the external network, just like the following code:

```go
 externalNetwork := util.GetExternalNetwork(newEip.Spec.ExternalSubnet)
```

if you use the field `newEip.Spec.ExternalSubnet` directly, this will cause compatibility issues. becasue the field `newEip.Spec.ExternalSubnet` may be empty.

this is not a great solution, but it works. if you have a better solution, please let me know.

**Review by bobz965 — APPROVED** [review]

File: pkg/util/net.go [hunk 1]
```diff
@@ -26,6 +26,8 @@ const (
 	IPv6Loopback         = "::1/128"
 	IPv6Multicast        = "ff00::/8"
 	IPv6LinkLocalUnicast = "FE80::/10"
+
+	VpcExternalNet = "ovn-vpc-external-network"
```
Comment (hongzhen-ma) [inline pkg/util/net.go:30]: 这个是不是还是放在 const 文件比较合适？其他vpc 相关的定义，都在const 文件里
net 文件里的const取值，只有 IP校验相关的地址

**Review by hongzhen-ma — APPROVED** [review]

File: pkg/util/net.go (hunk 1 above)
Comment (shane965) [reply pkg/util/net.go:30]: 嗯嗯，如果作为一个常量放在这里确实不太合适, `VpcExternalNet = "ovn-vpc-external-network"`的这个值不应该再给到外部引用，直接改成下面这种方式可能更好
```
file: pkg/util/net.go
var vpcExternalNet = "ovn-vpc-external-network"
```

Comment (hongzhen-ma) [reply pkg/util/net.go:30]: 如果考虑跟之前实现的兼容性，就还得保持 const 的定义，变量应该不行

原来使用方式的文档
https://kubeovn.github.io/docs/v1.12.x/guide/eip-snat/

**Review by hongzhen-ma — APPROVED** [review]

**Pull request merged**